File: Java/Ex1.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class Ex1 {
    public TreeNode inorderNext(TreeNode root, TreeNode p) {
        TreeNode next = null;
        TreeNode current = root;
        while (current != null) {
            if (p.val < current.val) {
                next = current;
                current = current.left;
            } else {
                current = current.right;
            }
        }
        return next;
    }

    public static void main(String[] args) {
        Ex1 tree = new Ex1();
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(1);
        root.right = new TreeNode(4);
        TreeNode p = root.left; // Node with value 1

        TreeNode nextNode = tree.inorderNext(root, p);
        System.out.println(nextNode != null ? nextNode.val : "null");
    }
}

File: Python/ex1.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorderNext(root, p):
    def leftmost(node):
        while node.left:
            node = node.left
        return node

    # Case 1: right child exists
    if p.right:
        return leftmost(p.right)

    # Case 2: no right child, find the first ancestor for which p is in the left subtree
    ancestor, current = None, root
    while current != p:
        if p.val < current.val:
            ancestor, current = current, current.left
        else:
            current = current.right
    return ancestor

# Example binary tree and test node
root = TreeNode(3)
root.left = TreeNode(1)
root.right = TreeNode(4)
p = root.left  # Node with value 1

print(inorderNext(root, p).val if inorderNext(root, p) else "null")
